Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/platform

cdumez opened the pull request:
#### 5dc8660948b7bc86ff19b9b09e55687fcc533aff
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/platform
<a href="https://bugs.webkit.org/show_bug.cgi?id=283863">https://bugs.webkit.org/show_bug.cgi?id=283863</a>

Reviewed by Sam Weinig.

* Source/WebCore/platform/sql/SQLiteStatement.cpp:
(WebCore::SQLiteStatement::bindBlob):
(WebCore::SQLiteStatement::columnValue):
(WebCore::SQLiteStatement::columnText):
(WebCore::SQLiteStatement::columnBlobAsString):
(WebCore::SQLiteStatement::columnBlobAsSpan):
* Source/WebCore/platform/text/DateTimeFormat.cpp:
* Source/WebCore/platform/text/PlatformLocale.cpp:
* Source/WebCore/platform/text/PlatformLocale.h:
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:

Canonical link: <a href="https://commits.webkit.org/287199@main">https://commits.webkit.org/287199@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14df8d80056b54914ce5ec50a9cafab3003d6f1a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32145 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83423 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30025 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66959 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6088 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61686 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19608 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71322 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41993 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49068 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25804 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28365 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70179 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26207 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84792 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6128 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69912 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6290 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67706 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69167 "Found 9 new API test failures: /TestWebKit:WebKit.DownloadDecideDestinationCrash, /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /TestWebKit:WebKit2UserMessageRoundTripTest.WKURLRequestRef, /TestWebKit:WebKit.InjectedBundleFrameHitTest, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit.GeolocationBasicWithHighAccuracy, /TestWebKit:WebKit.InjectedBundleBasic, /TestWebKit:WebKit.ParentFrame, /TestWebKit:WebKit.GeolocationBasic (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13206 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11889 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12156 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6073 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/11976 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6058 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9495 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7847 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -133,13 +131,14 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
-    std::array<float, 10> numLengths = { };
-    for (int i = 0; i < 10; i++)
+    std::array<float, 10> numLengths;
+    for (int i = 0; i < 10; ++i)
```
Can indexedRange() be used here? (I ask because I am not clear on how well it handles mutable references). (Also, all this code is bananas).

cdumez replied:
I actually don't know if indexedRange() handles mutable references well.

weinig commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -133,13 +131,14 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
-    int widestNumOneThroughFour = std::distance(numLengths.data(), std::max_element(numLengths.data() + 1, numLengths.data() + 5));
-    int widestNumNonZero = std::distance(numLengths.data(), std::max_element(numLengths.data() + 1, numLengths.data() + 10));
+    std::span numLengthsSpan { numLengths };
+    int widestNum = std::distance(numLengthsSpan.begin(), std::max_element(numLengthsSpan.begin(), numLengthsSpan.end()));
```
Can you use std::ranges for these? e.g. `std::ranges::max_element(numLengthsSpan))` rather than `std::max_element(numLengthsSpan.begin(), numLengthsSpan.end()))`?

weinig commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -133,13 +131,14 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
-    for (int i = 0; i < 10; i++)
+    std::array<float, 10> numLengths;
+    for (int i = 0; i < 10; ++i)
         numLengths[i] = measurer.measureText([NSString stringWithFormat:@"%d", i]);
```
Not for this change, but there is no reason for these to be going through NSString. measureText() takes a WTF::String.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/platform
https://bugs.webkit.org/show_bug.cgi?id=283863

Reviewed by Sam Weinig.

* Source/WebCore/platform/sql/SQLiteStatement.cpp:
(WebCore::SQLiteStatement::bindBlob):
(WebCore::SQLiteStatement::columnValue):
(WebCore::SQLiteStatement::columnText):
(WebCore::SQLiteStatement::columnBlobAsString):
(WebCore::SQLiteStatement::columnBlobAsSpan):
* Source/WebCore/platform/text/DateTimeFormat.cpp:
* Source/WebCore/platform/text/PlatformLocale.cpp:
* Source/WebCore/platform/text/PlatformLocale.h:
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:

Canonical link: https://commits.webkit.org/287199@main
